### ISSUE TITLE
collection_from_enum I18n unusable with has_many nested forms

### ISSUE THREAD

**Issue by fabien-michel** [formtastic/formtastic#1151]
the `collection_from_enum` method try to fetch translation with the following key: 

``` ruby
key = "activerecord.attributes.#{object_name}.#{method_name.pluralize}.#{name}"
```

In case where the field is in a has_many nested (eg: with cocoon) form, object_name can be like this : 

``` ruby
car[wheels_attributes][0]
```

So it is impossible to use the translation.
I suggest to change object_name by `model_name.underscore` or something like this to obtain just the model name as expected to use in the translation yml file.

**Comment by justinfrench** [comment]
@fabien-michel can you have a go at working up a failing test and a patch?

**Comment by fabien-michel** [comment]
Hi,
I am a beginner in ruby and rails, so that I'm not sure of my code and I'd post it here instead of doing a PR. Also I have not found a patch that works.

Here a failing test in radio_input_spec.rb
```
describe 'for a nested enum column' do
    before do

      ::I18n.backend.store_translations :en, :activerecord => { :attributes => { :author => { :statuses => { active: "Translated_active", inactive: "Translated_inactive"} }}}

      @bob.stub(:status) { 'inactive' }
      statuses = ActiveSupport::HashWithIndifferentAccess.new("active"=>0, "inactive"=>1)
      @bob.class.stub(:statuses) { statuses }
      @bob.stub(:defined_enums) { { "status" => statuses } }
      @new_post.stub(:author).and_return(@bob)
    end


    before do
      concat(semantic_form_for(@new_post) do |builder|
          inputs = builder.inputs :for => [:author, @bob] do |bob_builder|
            concat(bob_builder.input(:status, :as => :radio))
          end
          concat(inputs)
        end)
    end


    it 'should translate the title when column is in a nested form' do
      output_buffer.should have_tag("label", /Translated_active/)
      output_buffer.should have_tag("label", /Translated_inactive/)
    end

    after do
      ::I18n.backend.store_translations :en, {}
    end
        
  end
```

But i fail to find a valid patch as my previous suggestion ```model_name.underscore``` make the ```MyInput#collection_from_enum when an enum is defined for the method with translations returns an Array of EnumOption objects based on the enum options hash``` test fail because ```model_name```method call ```object.class.name``` which crash because in this case object is a double.

**Comment by justinfrench** [comment]
@fabien-michel I've applied your new test in a new branch, and I get one failure, as expected:

```
Expected [large chunk of HTML] to have an element matching "label" with inner text /Translated_active/, but found 0
```

Next, I modified `collection_from_enum` to see what `key` looked like during the test with a `puts`:

```
        def collection_from_enum
          if collection_from_enum?
            method_name = method.to_s

            enum_options_hash = object.defined_enums[method_name]
            enum_options_hash.map do |name, value| 
              key = "activerecord.attributes.#{object_name}.#{method_name.pluralize}.#{name}"
              puts key.inspect
              label = ::I18n.translate(key, :default => name.humanize) 
              [label, name]
            end
          end
        end
```

The output during the test run is:

```
"activerecord.attributes.post[author_attributes].statuses.active"
"activerecord.attributes.post[author_attributes].statuses.inactive"
```

Looking at your test set-up, you're hoping to use this i18n data:

```
::I18n.backend.store_translations :en, :activerecord => { :attributes => { :author => { :statuses => { active: "Translated_active", inactive: "Translated_inactive"} }}}
```

So I guess the goal is to modify the `key` in collection_from_enum to:

```
"activerecord.attributes.author.statuses.(active|inactive)"
```

Is that right? I'm not expert on enum fields or i18n or nested forms, but why wouldn't it be nested?

```
"activerecord.attributes.post.author.statuses.(active|inactive)"
```

If you could confirm that I correctly understand the test case and what needs to change, I'll look into a patch.

**Comment by fabien-michel** [comment]
Thank you to look at it !
Yes, the goal is to fetch the translation for "activerecord.attributes.author.statuses.(active|inactive)" the same key when fetching translation for the field when it is not nested.
 
Using model_name should solve the problem, but it made another test fail in MyInput#collection_from_enum.
But I think this fail because the use of double(). At my advice it should not occur out of the test environment, but I'm not sure at all about that.

**Comment by justinfrench** [comment]
Why `activerecord.attributes.author.statuses.(active|inactive)` over `activerecord.attributes.post.author.statuses.(active|inactive)`?

**Comment by fabien-michel** [comment]
Hmm, I don't see why we could have use of specific translation for a nested model

**Comment by justinfrench** [comment]
Okay, I've been looking into this a bit more. My first step was checking what simple_form does. They have a "won't fix" issue plataformatec/simple_form#1063 which suggests that there are some issues in Rails for enums on user input (like forms).

I believe we _can_ patch Formtastic using `model_name.downcase` or a modification of `object_name` to suit your needs, but I'm not convinced that we _should_:

1. What _are_ the Rails team issues with Enum — will the implementation change?
2. Is the simple_form issue similar to the formtastic one, or different? How?
3. What is the _right_ i18n key? `post.author.statuses` or `author.statuses` or something else?
4. What _don't I know_ about this problem that might introduce a bug for others?

So, for now, I think my suggestion is to write your own monkey patch in an initializer. Something like this:

```ruby
# config/initializers/formtastic_i18n_patch.rb
module Formtastic::Inputs::Base::Collections
  def collection_from_enum
    if collection_from_enum?
      method_name = method.to_s
      enum_options_hash = object.defined_enums[method_name]
      enum_options_hash.map do |name, value| 
        # patched line below
        key = "activerecord.attributes.#{model_name.downcase}.#{method_name.pluralize}.#{name}"
        label = ::I18n.translate(key, :default => name.humanize) 
        [label, name]
      end
    end
  end
end
```

I'm not interested in merging a fix into master at this time though. Closing this ticket for now, but please feel free to re-open if you need a hand with the patch.

**Comment by sandfoxme** [comment]
plz reopen. i have a fix, PR incoming
